### PR TITLE
Fix spacing rules for functions

### LINKED ISSUE #5808: False positive LT01 with `COMMENT ON FUNCTION ... IS`
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff detects a false violation of LT01 with Postgres code in the form of
```sql
COMMENT ON FUNCTION x IS 'y';
```

If `fix` is applied, it will be changed to:
```sql
COMMENT ON FUNCTION xIS 'y';
```

### Expected Behaviour

No violation should be found.

### Observed Behaviour

Linting fails LT01

```
L:   1 | P:  22 | LT01 | Unexpected whitespace before 'IS' keyword.
                       | [layout.spacing]
```

### How to reproduce

`echo "COMMENT ON FUNCTION x IS 'y';" | sqlfluff lint --dialect postgres -`

### Dialect

postgres

### Version

3.0.3

### Configuration

None

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #5869: REBALANCE hint produces unexpected spacing error in spark SQL
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

REBALANCE hint produces unexpected spacing error in spark SQL
other hints (with brackets) pass as expected

### Expected Behaviour

Pass without error

### Observed Behaviour

linting failed with error 
```
== [stdin] FAIL
L:   2 | P:  21 | LT01 | Unexpected whitespace before end hint '*/'.
                       | [layout.spacing]
All Finished 📜 🎉!
```

### How to reproduce

`echo 'INSERT OVERWRITE TABLE aaa.bbb PARTITION (a)\nSELECT /*+ REBALANCE */ 1 AS a FROM xxx.yyy' | sqlfluff lint --dialect sparksql -`

### Dialect

sparksql

### Version

3.0.6

### Configuration

default settings

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #5978: tsql LT01 bug. `GO` statement must stay on its own line
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I tried to format a tsql file and `sqlfluff` broke the codebase

### Expected Behaviour

Code is formatted and can still be executed.

### Observed Behaviour

Cannot execute the code anymore.

### How to reproduce

create a file query:

```sql
DROP FUNCTION IF EXISTS INTERNAL_ETL.DIL_md_ScdTest
GO
DROP FUNCTION IF EXISTS INTERNAL_ETL.DIL_md_ScdTest
GO
```

Then run the following command:
```ps1
PS C:\Users\fmoeller\Desktop\RepositoryVorlage>  sqlfluff fix --dialect tsql  .\query.sql
==== finding fixable violations ====
== [query.sql] FAIL
L:   1 | P:  52 | LT01 | Unexpected line break. [layout.spacing]
L:   3 | P:  52 | LT01 | Unexpected line break. [layout.spacing]
== [query.sql] FIXED
2 fixable linting violations found
```

The output is 

```sql
DROP FUNCTION IF EXISTS INTERNAL_ETL.DIL_md_ScdTestGO
DROP FUNCTION IF EXISTS INTERNAL_ETL.DIL_md_ScdTestGO
```

`GO` or `GO <n>` must always be on its own line.

### Dialect

tsql

### Version

```ps1
PS C:\Users\fmoeller\Desktop\RepositoryVorlage> sqlfluff version -v
==== sqlfluff ====
sqlfluff:                3.0.7 python:                 3.11.9
implementation:        cpython verbosity:                   1
dialect:                  tsql templater:                 raw
rules:                                all
```

### Configuration

none

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by WittierDinosaur**
### Brief summary of the change made
Function names shouldn't have touch after, function parameters should have touch before - as functions can be bare.
Fixes #5808 
Fixes #5869
Fixes #5978

### Are there any other side effects of this change that we should be aware of?
Hopefully no knock on effects. I reckon now that we have the function_contents segment, a few rules can be rewritten in a more elegant way - but this PR is already crowded enough. It may be worth a minor release due to the sheer size of the code change.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/rules/std_rule_cases/LT01-functions.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+rule: LT01
+
+test_pass_bare_function:
+  pass: COMMENT ON FUNCTION x IS 'y';
+  configs:
+    core:
+      dialect: postgres
+
+test_fail_spaced_function:
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/LT01-functions.yml:9]:
```suggestion
test_fail_space_function:
```

File: test/fixtures/rules/std_rule_cases/LT01-functions.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+rule: LT01
+
+test_pass_bare_function:
+  pass: COMMENT ON FUNCTION x IS 'y';
+  configs:
+    core:
+      dialect: postgres
+
+test_fail_space_function:
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/LT01-functions.yml:9]:
```suggestion
test_fail_spaced_function:
```

File: test/fixtures/rules/std_rule_cases/LT01-functions.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+test_pass_bare_function:
+  pass: COMMENT ON FUNCTION x IS 'y';
+  configs:
+    core:
+      dialect: postgres
+
+test_fail_spaced_function:
+  fail_str: COMMENT ON FUNCTION x (foo) IS 'y';
+  fix_str: COMMENT ON FUNCTION x (foo) IS 'y';
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/LT01-functions.yml:11]:
```suggestion
  fix_str: COMMENT ON FUNCTION x(foo) IS 'y';
```

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 4]
```diff
@@ -962,6 +962,58 @@ class FunctionNameSegment(ansi.FunctionNameSegment):
...
+    """Datetime function contents segment."""
+
+    match_grammar = Sequence(
+        Bracketed(
+            Delimited(
+                Ref("DatetimeUnitSegment"),
+                Ref("DatePartWeekSegment"),
+                Ref(
+                    "FunctionContentsGrammar",
+                ),
```
Comment (keraion) [inline src/sqlfluff/dialects/dialect_bigquery.py:975]: nit:
```suggestion
                Ref("FunctionContentsGrammar"),
```

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 5]
```diff
@@ -1396,6 +1396,43 @@ class FunctionNameSegment(BaseSegment):
     )
 
 
+class DateTimeFunctionContentsSegment(BaseSegment):
+    """Datetime function contents."""
+
+    type = "function_contents"
+
+    match_grammar = Sequence(
```
[the remark this answers is not in the document]
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:1404]: Interesting quirk that we should probably fix. You cannot start a match grammar with Bracketed(). The sequence wrapping allows it to work, OneOf, AnyOf, etc would also work. It you remove the Sequence here, it breaks

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 6]
```diff
@@ -1396,6 +1396,43 @@ class FunctionNameSegment(BaseSegment):
...
+        Bracketed(
+            Delimited(
+                Ref("DatetimeUnitSegment"),
+                Ref(
+                    "FunctionContentsGrammar",
+                    # The brackets might be empty for some functions...
+                    optional=True,
+                ),
+            ),
+            # parse_mode=ParseMode.GREEDY,
```
[the remark this answers is not in the document]
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:1414]: agreed on these, will sort tomorrow

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (keraion) [reply src/sqlfluff/dialects/dialect_ansi.py:1404]: Oh, that is interesting.

**Review by keraion — APPROVED** [review]
After the change this LGTM!

**Pull request merged**